src/geometry.cpp

```diff
@@ -536,6 +536,196 @@ extern "C" double openmc_calculate_optical_thickness(
+    Position r_i = voxel_box.sample(&seed);
+    Position r_j = voxel_box.sample(&seed);
+
+    const Position directions[6] = {
+      {1.0, 0.0, 0.0},  // +x direction
+      {-1.0, 0.0, 0.0}, // -x direction
+      {0.0, 1.0, 0.0},  // +y direction
+      {0.0, -1.0, 0.0}, // -y direction
+      {0.0, 0.0, 1.0},  // +z direction
+      {0.0, 0.0, -1.0}  // -z direction
+    };
+
+    for (int dir = 0; dir < 6; dir++) {
+      Position b;
+
+      switch (dir) {
+      case 0: // +x direction
+        b = {voxel_max.x, r_j.y, r_j.z};
+        break;
+      case 1: // -x direction
+        b = {voxel_min.x, r_j.y, r_j.z};
+        break;
+      case 2: // +y direction
+        b = {r_j.x, voxel_max.y, r_j.z};
+        break;
+      case 3: // -y direction
+        b = {r_j.x, voxel_min.y, r_j.z};
+        break;
+      case 4: // +z direction
+        b = {r_j.x, r_j.y, voxel_max.z};
+        break;
+      case 5: // -z direction
+        b = {r_j.x, r_j.y, voxel_min.z};
+        break;
+      }
+
+      // Calculate optical thickness from r_i to b
+      double optical_thickness = openmc_calculate_optical_thickness(r_i, b);
+
+      // Transport operator: exp(-tau)/|b-r_i|^2
+      double transport =
+        std::exp(-optical_thickness) / ((b - r_i).norm() * (b - r_i).norm());
+      total_transport += transport;
+      valid_samples++;
+    }
+  }
+
+  if (valid_samples > 0) {
+    *self_transport_output = total_transport / valid_samples;
+  } else {
+    *self_transport_output = 0.0;
+  }
+}
+
+extern "C" void openmc_calculate_g_ij(Position start_voxel_min,
+  Position start_voxel_max, Position end_voxel_min, Position end_voxel_max,
+  int num_rays, double* g_ij_output)
+{
+  double transport_operator;
+  openmc_get_transport_operator_between_voxels(start_voxel_min, start_voxel_max,
+    end_voxel_min, end_voxel_max, num_rays, &transport_operator);
+
+  // note that v_i is not actually used in the calculation of g_ij
+  double v_i = (start_voxel_max.x - start_voxel_min.x) *
+               (start_voxel_max.y - start_voxel_min.y) *
+               (start_voxel_max.z - start_voxel_min.z);
+
+  double v_j = (end_voxel_max.x - end_voxel_min.x) *
+               (end_voxel_max.y - end_voxel_min.y) *
+               (end_voxel_max.z - end_voxel_min.z);
+
+  // g_ij = transport_operator * (1 / (4 pi V_i)) * V_i * V_j
+  // = transport operator * (1 / (4 pi)) * V_j
+  *g_ij_output = transport_operator * (1.0 / (4.0 * PI)) * v_j;
+}
+
+extern "C" void openmc_calculate_g_ii(
+  Position voxel_min, Position voxel_max, int num_rays, double* g_ii_output)
+{
+  double self_transport;
+  openmc_get_self_transport_operator(
+    voxel_min, voxel_max, num_rays, &self_transport);
+
+  double volume = (voxel_max.x - voxel_min.x) * (voxel_max.y - voxel_min.y) *
+                  (voxel_max.z - voxel_min.z);
+
+  // NOTE: assume cube for area calculation
+  double area = (voxel_max.x - voxel_min.x) * (voxel_max.y - voxel_min.y);
```

Comment:
No need to assume a cube...
```suggestion
  double diagonal = voxel_max - voxel_min;
  double volume = diagonal.x * diagonal.y * diagonal*z;
  double area = 2 * ( diagonal.x * diagonal.y + diagonal.x * diagonal.z + diagonal.y * diagonal.z);
```

Reply:
Nice refactor! I am not sure why I assumed it would be area of only one face. Total surface area makes more sense. After implementing, I am running into frequent issues of diagonal elements being negative. Investigating further

Reply:
Actually, this might simply be an undersampling issue. Increasing rays seems to fix this issue. I wonder if guardrails should be installed to allow for undersampling (e.g. ensuring g_ii < 1 so the P_ii values are never negative)